File: backend/schemas/artifact_jobs.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
)

from schemas.artifact_composer import (
    ArtifactComposeRequest,
    ArtifactComposeResponse,
)


ArtifactJobStatus = Literal[
    "queued",
    "running",
    "succeeded",
    "failed",
]


class ArtifactJobCreateRequest(
    ArtifactComposeRequest
):
    """
    Request accepted by the asynchronous
    prompt-to-artifact generation endpoint.
    """

    model_config = ConfigDict(
        extra="forbid",
        str_strip_whitespace=True,
    )


class ArtifactJobCreateResponse(BaseModel):
    """
    Initial response returned immediately
    after a background job is accepted.
    """

    model_config = ConfigDict(
        extra="forbid",
    )

    job_id: str = Field(
        min_length=32,
        max_length=32,
    )

    status: ArtifactJobStatus

    access_token: str = Field(
        min_length=32,
        max_length=256,
    )

    created_at: datetime
    expires_at: datetime

    status_url: str

    message: str = (
        "Artifact generation job accepted."
    )


class ArtifactJobStatusResponse(BaseModel):
    """
    Current background artifact job status.
    """

    model_config = ConfigDict(
        extra="forbid",
    )

    job_id: str = Field(
        min_length=32,
        max_length=32,
    )

    status: ArtifactJobStatus

    progress_percent: int = Field(
        ge=0,
        le=100,
    )

    stage: str = Field(
        min_length=1,
        max_length=160,
    )

    created_at: datetime
    updated_at: datetime
    expires_at: datetime

    artifact: (
        ArtifactComposeResponse
        | None
    ) = None

    error: str | None = Field(
        default=None,
        max_length=5_000,
    )


class ArtifactJobDeleteResponse(BaseModel):
    """
    Response returned after a completed or
    failed background job record is removed.
    """

    model_config = ConfigDict(
        extra="forbid",
    )

    job_id: str = Field(
        min_length=32,
        max_length=32,
    )

    deleted: bool

File: frontend/src/types/artifact-jobs.ts
import type {
  ArtifactComposeRequest,
  ArtifactComposeResponse,
} from './artifact-composer'

export type ArtifactJobStatus =
  | 'queued'
  | 'running'
  | 'succeeded'
  | 'failed'

export interface ArtifactJobCreateRequest
  extends ArtifactComposeRequest {}

export interface ArtifactJobCreateResponse {
  job_id: string
  status: ArtifactJobStatus
  access_token: string
  created_at: string
  expires_at: string
  status_url: string
  message: string
}

export interface ArtifactJobStatusResponse {
  job_id: string
  status: ArtifactJobStatus
  progress_percent: number
  stage: string
  created_at: string
  updated_at: string
  expires_at: string
  artifact:
    | ArtifactComposeResponse
    | null
  error: string | null
}

export interface ArtifactJobDeleteResponse {
  job_id: string
  deleted: boolean
}

export interface ActiveArtifactJob {
  jobId: string
  accessToken: string
  status: ArtifactJobStatus
  progressPercent: number
  stage: string
  createdAt: string
  expiresAt: string
  artifact:
    | ArtifactComposeResponse
    | null
  error: string | null
}